esp-hal/src/lib.rs

```diff
@@ -572,6 +572,115 @@ mod critical_section_impl {
     }
 }
 
+// The state of a re-entrant lock
+pub(crate) struct LockState {
+    #[cfg(multi_core)]
+    core: portable_atomic::AtomicU8,
+}
+
+impl LockState {
+    pub const fn new() -> Self {
+        Self {
+            #[cfg(multi_core)]
+            core: portable_atomic::AtomicU8::new(0),
+        }
+    }
+
+    #[cfg(multi_core)]
+    fn current_core() -> u8 {
+        // Note: 0 means unlocked.
+        match get_core() {
+            Cpu::ProCpu => 1,
+            Cpu::AppCpu => 2,
+        }
+    }
+}
+
+// This is preferred over critical-section as this allows you to have multiple
+// locks active at the same time rather than using the global mutex that is
+// critical-section.
+#[allow(unused_variables)]
+pub(crate) fn lock<T>(state: &LockState, f: impl FnOnce() -> T) -> T {
+    // In regards to disabling interrupts, we only need to disable
+    // the interrupts that may only eb calling this function.
+
+    #[cfg(not(multi_core))]
+    {
+        // Disabling interrupts is enough on single core chips to ensure mutual
+        // exclusion.
+
+        #[cfg(riscv)]
+        return riscv::interrupt::free(f);
+        #[cfg(xtensa)]
+        return xtensa_lx::interrupt::free(|_| f());
+    }
+
+    #[cfg(multi_core)]
+    {
+        use portable_atomic::Ordering;
+
+        let current_core = LockState::current_core();
+
+        // Check for re-entry
+        {
+            // Used Relaxed ordering since an Acquire has already been performed for the
+            // re-entrant case that is being checked here.
+            let locked_core = state.core.load(Ordering::Relaxed);
+            if locked_core == current_core {
+                // Note: Since this is re-entrant, interrupts have already been disabled.
+                return f();
+            }
+        }
+
+        let mut f = Some(f);
+
+        loop {
+            // Relaxed ordering is fine here since it's just a hint for the CEX below.
+            let locked_core = state.core.load(Ordering::Relaxed);
+
+            if locked_core != 0 {
+                // Re-entry should've been caught before the loop.
+                debug_assert_ne!(locked_core, current_core);
+
+                // Consider using core::hint::spin_loop(); Might need SW_INT.
+                continue;
+            }
+
+            // The lock is released.
+
+            let mut f = || {
+                // Use Acquire ordering in success to ensure `f()` "happens after" the lock is
+                // taken. Use Relaxed ordering in failure as there's no
+                // synchronisation happening.
+                if state
+                    .core
+                    .compare_exchange(0, current_core, Ordering::Acquire, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    let result = f.take().unwrap()();
```

Comment:
Please avoid `()()`. Also I'd prefer not having the unwrap, even though it's panic is unreachable. Just don't do anything on None.

Reply:
I went with your suggested approach below instead.